layout/plan.c

```diff
@@ -192,31 +256,37 @@ M0_INTERNAL void m0_layout_plan_fini(struct m0_layout_plan *plan)
 M0_INTERNAL int m0_layout_plan_get(struct m0_layout_plan *plan, uint64_t colour,
 				   struct m0_layout_plop **plop)
 {
-	if (plan == NULL || plop == NULL)
-		return M0_ERR(-EINVAL);
+	M0_PRE(plan != NULL);
+	M0_PRE(plop != NULL);
+	M0_PRE(plan->lp_op != NULL);
 
-	if (plan->lp_op == NULL)
-		return M0_ERR_INFO(-EINVAL, "plan %p is not built", plan);
+	m0_mutex_lock(&plan->lp_lock);
 
 	if (plan->lp_last_plop == NULL)
-		*plop = pplops_tlist_tail(&plan->lp_plops);
+		*plop = pplops_tlist_head(&plan->lp_plops);
 	else
-		*plop = pplops_tlist_prev(&plan->lp_plops, plan->lp_last_plop);
+		*plop = pplops_tlist_next(&plan->lp_plops, plan->lp_last_plop);
 
 	if (*plop == NULL) {
 		*plop = plop_alloc_init(plan, M0_LAT_DONE, NULL);
-		if (*plop == NULL)
+		if (*plop == NULL) {
+			m0_mutex_unlock(&plan->lp_lock);
```

Comment:
Yes, I think LAT_DONE should depend on all M0_LAT_OUT_READ plops. Then you can construct LAT_DONE in m0_layout_plan_build() and avoid special case in plan_get().

Reply:
Sounds good, done.